app_usb_aud_xk_316_mc_multirate_i2s/src/extensions/extra_i2s.xc

```diff
@@ -0,0 +1,364 @@
+    int rdPtr;
+    int size;
+    int fill;
+} fifo_t;
+
+static void init_fifo(fifo_t &f, int array[size], int size)
+{
+    f.wrPtr = size/2;
+    f.rdPtr = 0;
+    f.size = size;
+    f.fill = size/2;
+
+    unsafe
+    {
+        int * unsafe arrayPtr = &array[0];
+        memset(arrayPtr, 0, size * (sizeof(array[0])));
+    }
+}
+
+static inline unsigned fifo_pop(fifo_t &f, int array[], int &sample)
+{
+    /* Check for FIFO empty */
+    if (!f.fill)
+    {
+        sample = 0;
+        return 1;
+    }
+
+    sample = array[f.rdPtr];
+    f.rdPtr++;
+    f.fill--;
+
+    /* Check for wrap */
+    if (f.rdPtr >= f.size)
+    {
+        f.rdPtr = 0;
+    }
+
+    return 0;
+}
+
+static inline unsigned fifo_push(fifo_t &f, int array[], int sample)
+{
+    /* Check for FIFO full */
+    if(f.fill >= f.size)
+    {
+        return 1;
+    }
+
+    array[f.wrPtr] = sample;
+    f.wrPtr++;
+    f.fill++;
+
+    /* Check for wrap */
+    if(f.wrPtr >=f.size)
+    {
+        f.wrPtr = 0;
+    }
+
+    return 0;
+}
+
+#pragma unsafe arrays
+void UserBufferManagement(unsigned sampsFromUsbToAudio[], unsigned sampsFromAudioToUsb[])
+{
+    unsafe
+    {
+        outuint((chanend) uc_i2s, 1);
+
+        for(size_t i = 0; i < EXTRA_I2S_CHAN_COUNT_OUT; i++)
+        {
+            outuint((chanend)uc_i2s, sampsFromUsbToAudio[i + EXTRA_I2S_CHAN_INDEX_OUT]);
+        }
+        outct((chanend)uc_i2s, XS1_CT_END);
+
+        for(size_t i = 0; i< EXTRA_I2S_CHAN_COUNT_IN; i++)
+        {
+            sampsFromAudioToUsb[i + EXTRA_I2S_CHAN_INDEX_IN] = inuint((chanend) uc_i2s);
+        }
+        chkct((chanend)uc_i2s, XS1_CT_END);
+    }
+}
+
+static inline void trigger_src(streaming chanend c_dsp[SSRC_N_INSTANCES],
+                                int srcInputBuff[SSRC_N_INSTANCES][SSRC_N_IN_SAMPLES][SSRC_CHANNELS_PER_INSTANCE],
+                                fifo_t &fifo,
+                                int srcOutputBuff[SSRC_N_CHANNELS * SSRC_OUT_BUFF_SIZE * 2])
+{
+    int nSamps = 0;
+#pragma loop unroll
+    for (int i=0; i<SSRC_N_INSTANCES; i++)
+    {
+#pragma loop unroll
+        for (int j=0; j<SSRC_N_IN_SAMPLES; j++)
+        {
+#pragma loop unroll
+            for (int k=0; k<SSRC_CHANNELS_PER_INSTANCE; k++)
+            {
+                c_dsp[i] <: srcInputBuff[i][j][k];
+            }
+        }
+    }
+
+    /* Get number of samples to receive from all SRC cores */
+    /* Note, all nSamps should be equal */
+#pragma loop unroll
+    for (int i=0; i < SSRC_N_INSTANCES; i++)
+    {
+        c_dsp[i] :> nSamps;
+    }
+
+#pragma loop unroll
+    for (int i=0; i<SSRC_N_INSTANCES; i++)
+    {
+        unsigned error = 0;
+        for (int j=0; j < nSamps; j++)
+        {
+#pragma loop unroll
+            for (int k=0; k<SSRC_CHANNELS_PER_INSTANCE; k++)
+            {
+                int sample;
+                c_dsp[i] :> sample;
+                unsafe
+                {
+                    error &= fifo_push(fifo, srcOutputBuff, sample);
```

Comment:
Should this be `|=`? Otherwise subsequent returns of zero inside the loop would "unset" error.

Reply:
Yes, thanks, I had "success" returned previously and forgot to update this..